### PR TITLE
callCabal2nix: Don't rebuild unchanged cabal file (Fixed #23880)

### PR DESCRIPTION
**Pull request by ElvishJerricco**
###### Motivation for this change

#23880 caused #24245 and was reverted. This PR fixes the issue by only applying `filterSource` on sources of type `path`. I'm usually weary of these dynamically typed things, but I think it's worth it in this case. Editing haskell files should not result in the cabal file being reevaluated by `cabal2nix`, and it definitely shouldn't cause the entire source tree to be copied into the nix store.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by PierreR** [comment]
@ElvishJerricco what should I do when I do want to rebuild anytime I change the `rev` in 
```
          package_latest = hsuper.callCabal2nix "language-puppet" (pkgs.fetchFromGitHub {
            owner  = "xxxxx";    repo   = "language-puppet";
            rev = "2845454bd0bc95771d875df273adc63e24ddaded";
            sha256 = "1iz5gflicris9jf5fvzfv8i69hmsr432dbv4hrpvqk0z1sqh0qbb";
          }) { };
```
This is actually quite a common workflow for me right now. I want to get the latest revision of a package but the cabal file is left unchanged because it is not released yet.

Is there an easy way to do that without removing the cached derivation ?

**Comment by ElvishJerricco** [comment]
@PierreR That should work given this change no problem. Notice that it overrides the src field to put back the src argument; if that src is different, it will yield a different derivation, even if the cabal file is unchanged.

**Comment by PierreR** [comment]
@ElvishJerricco I am using `nix-shell -A project release.nix` with [release.nix](http://nixpaste.lbr.uno/ONtawvJk)

I was using a simple version of this before with an older pinned nixpkgs version and it did refresh. If you don't see anything that is obviously wrong in 'release.nix' I will check that with older version to be sure.

Thanks for your quick feedback.

**Comment by ElvishJerricco** [comment]
@PierreR `fetchFromGitHub` is a fixed output derivation. If you don't change the `sha256` argument, then any changes to the other arguments will not trigger a rebuild, even if such a rebuild would fail since the `src` hash is wrong. Notice that your nixpaste link contains a `fetchFromGitHub` expression that does not build. To test this, you can use `--check`, like `nix-build release.nix -A language-puppet_latest.src --check`, if you add `language-puppet_latest` to the `release.nix` output. The reason that you did not notice this is that you kept the `sha256` from a different version that you had successfully built, which told Nix not to retry the src download, and just use the old one.

I think `fetchFromGitHub` used to not exhibit this behavior, and would get treated as a different derivation if only the rev changed and not the sha (which would fail), so that might be why this used to work for you.

**TL;DR:** This is `fetchFromGitHub`'s fault, not `callCabal2nix`'s. You need to update your `sha256`.

**Comment by vcunat** [comment]
The typical hack is to make some random change in the hash.

**Comment by PierreR** [comment]
@ElvishJerricco I wasn't aware about the change. Thanks for your valuable help !
